tools/clang/lib/SPIRV/DeclResultIdMapper.cpp

```diff
@@ -1016,7 +1016,22 @@ std::vector<SpirvVariable *> DeclResultIdMapper::collectStageVars() const {
   llvm::DenseSet<SpirvInstruction *> seenVars;
   for (const auto &var : stageVars) {
     auto *instr = var.getSpirvInstr();
-    if (seenVars.count(instr) == 0) {
+    spv::StorageClass sc = var.getStorageClass();
+    bool validStorageClassForVar = true;
+    switch (sc) {
+    // Non input raytracing stage variables should be ignored since
+    // they don't participate in any interface matching
```

Comment:
Q) Is this for skipping over raytracing builtin variables?

A) No these are for interface variables.
Maybe you can take a look at example at end of spec : https://github.com/KhronosGroup/GLSL/blob/master/extensions/nv/GLSL_NV_ray_tracing.txt

rayPayloadNV/rayPayloadInNV/hitAttributeNV/callableDataNV/callableDataInNV are all storage classes for interface variables. 

For ex rayPayloadNV created in a ray-generation shader should match rayPayloadInNV in the closest-hit shader/any-hit shader invoked or results are undefined
A ray tracing pipeline can contain large number of shaders(for ex hit shader for each geometry/material )and we do not verify interfaces are matching due to combinatorial explosion

So technically these are stage variables and not really resourceVars/builtinVars.

I am not of strong opinion either way, so i can look into not treating them as StageVars if you feel that is the right way to go about it

Thanks!

Reply:
@ehsannas  : Seems you might have missed my comment above
Any thoughts on next steps? Thanks

Reply:
@alelenv , thanks for pointing out. I missed your comment here.

By skipping over variables with these storage classes in `DeclResultIdMapper::collectStageVars()`, you effectively won't have them in your `OpEntryPoint` instruction. Is this your expected behavior?

Reply:
Yes. We have not amended the OpEntryPoint clause : 
"Interface is a list of <id> of global OpVariable instructions with either Input or Output for its Storage Class operand."

Reply:
Ah, ok. This was part of the reason for my confusion. It's more clear now. Thanks for the explanation.

Regarding adding these variables to the `stageVars` vector:
I'm concerned about whether the usages of `stageVars` fit your needs.

For example:
1) In the `createRayTracingNVStageVar` function you create a stageVar for which the SigPoint is `nullptr` and add it to `stageVars` vector.
In `DeclResultIdMapper::checkSemanticDuplication`, we iterate over `stageVars` and access the object's SigPoint to figure out whether it's for input or output. Theoretically we will hit a segfault there for the raytracing stage vars?
(I'm wondering why we haven't seen this in the segfault in the tests)

2) Similarly, we iterate over all `stageVars` in `DeclResultIdMapper::finalizeStageIOLocations` to assign `Location` decoration to stage variables. Do you need this logic to run for the newly-created raytracing stage variables?

Reply:
> I'm wondering why we haven't seen this in the segfault in the tests

That's because either `var.getSemanticStr()` is empty or `var.isSpirvBuitin()` is `true`. The latter I believe. So we skip over them in the for loop.

Reply:
Its probably because we never invoke checkSemanticDuplication since i back out in decorateStageIOLocations for raytracing shaders

It does seem like i am using StageVars incorrectly for my use case and doing early returns in most places. Let me look into this

Reply:
Yeah, Sounds good. Feel free to add a different vector to declResultIdMapper (e.g. `stageVarsRayTracingNV`) and work with that.